File: src/main/java/net/reflact/client/network/ClientNetworkManager.java
package net.reflact.client.network;

import com.google.gson.Gson;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.reflact.client.ClientData;
import net.reflact.common.network.PacketRegistry;
import net.reflact.common.network.packet.CastSpellPacket;
import net.reflact.common.network.packet.ManaUpdatePacket;
import net.reflact.common.network.packet.ReflactPacket;
import net.reflact.common.network.packet.S2CSyncItemPacket;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ClientNetworkManager {
    private static final Logger LOGGER = LoggerFactory.getLogger("ReflactNet");
    private static final Gson gson = new Gson();

    public static void init() {
        ClientPlayNetworking.registerGlobalReceiver(ReflactJsonPayload.ID, (payload, context) -> {
            String data = payload.data();
            int splitIndex = data.indexOf(":");
            if (splitIndex == -1) return;
            
            String id = data.substring(0, splitIndex);
            String json = data.substring(splitIndex + 1);
            
            context.client().execute(() -> handlePacket(id, json));
        });
    }

    private static void handlePacket(String id, String json) {
        Class<? extends ReflactPacket> clazz = PacketRegistry.get(id);
        if (clazz == null) {
            LOGGER.warn("Unknown packet ID: {}", id);
            return;
        }
        
        try {
            ReflactPacket packet = gson.fromJson(json, clazz);
            processPacket(packet);
        } catch (Exception e) {
            LOGGER.error("Failed to process packet {}: {}", id, e.getMessage());
        }
    }

    private static void processPacket(ReflactPacket packet) {
        if (packet instanceof ManaUpdatePacket manaPacket) {
            ClientData.currentMana = manaPacket.currentMana();
        } else if (packet instanceof S2CSyncItemPacket syncPacket) {
            net.reflact.client.managers.ClientItemManager.cacheItem(syncPacket.item());
        }
    }

    public static void sendPacket(ReflactPacket packet) {
        String id = PacketRegistry.getId(packet.getClass());
        if (id == null) {
            LOGGER.warn("Tried to send unregistered packet: {}", packet.getClass().getName());
            return;
        }
        String data = id + ":" + gson.toJson(packet);
        if (ClientPlayNetworking.canSend(ReflactJsonPayload.ID)) {
            ClientPlayNetworking.send(new ReflactJsonPayload(data));
        }
    }
}

File: src/main/kotlin/net/reflact/client/network/ClientNetworkManager.kt
package net.reflact.client.network

import com.google.common.io.ByteStreams
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.reflact.client.ClientData
import net.reflact.client.render.DamageIndicatorRenderer
import net.reflact.common.network.PacketRegistry
import net.reflact.common.network.packet.DamageIndicatorPacket
import net.reflact.common.network.packet.ManaUpdatePacket
import net.reflact.common.network.packet.MapDataPacket
import net.reflact.common.network.packet.ReflactPacket
import net.reflact.common.network.packet.S2CSyncItemPacket
import org.slf4j.LoggerFactory
import java.io.ByteArrayInputStream
import java.io.DataInputStream

object ClientNetworkManager {
    private val LOGGER = LoggerFactory.getLogger("ReflactNet")

    fun init() {
        ClientPlayNetworking.registerGlobalReceiver(ReflactJsonPayload.ID) { payload, context ->
            // ReflactJsonPayload now contains raw bytes despite the name
            val data = payload.dataBytes

            context.client().execute {
                try {
                    val input = DataInputStream(ByteArrayInputStream(data))
                    val id = input.readUTF()

                    val clazz = PacketRegistry.get(id)
                    if (clazz == null) {
                        LOGGER.warn("Unknown packet ID: {}", id)
                        return@execute
                    }

                    // Instantiate via no-arg constructor
                    val packet = clazz.getDeclaredConstructor().newInstance()
                    packet.decode(input)

                    processPacket(packet)

                } catch (e: Exception) {
                    LOGGER.error("Failed to decode packet", e)
                }
            }
        }
    }

    private fun processPacket(packet: ReflactPacket) {
        if (packet is ManaUpdatePacket) {
            ClientData.currentMana = packet.currentMana
            ClientData.maxMana = packet.maxMana
        } else if (packet is S2CSyncItemPacket) {
            if (packet.item != null) {
                net.reflact.client.managers.ClientItemManager.cacheItem(packet.item!!)
            }
        } else if (packet is MapDataPacket) {
            net.reflact.client.map.MapTextureManager.INSTANCE.setMapData(
                packet.colors,
                packet.width,
                packet.height
            )
        } else if (packet is DamageIndicatorPacket) {
            DamageIndicatorRenderer.add(packet.entityId, packet.amount, packet.isCritical)
        }
    }

    fun sendPacket(packet: ReflactPacket) {
        val id = PacketRegistry.getId(packet.javaClass)
        if (id == null) {
            LOGGER.warn("Tried to send unregistered packet: {}", packet.javaClass.name)
            return
        }

        try {
            val out = ByteStreams.newDataOutput()
            out.writeUTF(id)
            packet.encode(out)

            if (ClientPlayNetworking.canSend(ReflactJsonPayload.ID)) {
                ClientPlayNetworking.send(ReflactJsonPayload(out.toByteArray()))
            }
        } catch (e: Exception) {
            LOGGER.error("Failed to send packet", e)
        }
    }
}
